### PR TITLE
WIP: Unix socket builder

### PR DESCRIPTION
**Pull request by nicoschoenmaker**
Still work in progress. Seems to be working. Interested in feedback!

Basically removed the existing runner classes, added a `RunnerInterface` instead:
```php
interface RunnerInterface
{
    public function execute(string $type, ContentItem $item): string;
}
```

With two implementations: `SingleProcessRunner` (current behaviour) and `UnixSocketRunner` (very fast compiles because node/typescript/less/etc. is already in memory)

To do:
- More unit-tests on the PHP side
- Make the `/tmp/build` path dynamic. Thinking about putting it in the cache dir

### CODE REVIEW

File: src/Bundler/Runner/Exception/SocketException.php [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?php
+/**
+ * @copyright 2017 Hostnet B.V.
+ */
+declare(strict_types=1);
+
+namespace Hostnet\Component\Resolver\Bundler\Runner\Exception;
+
+class SocketException extends \Exception
```
Comment (yannickl88) [inline src/Bundler/Runner/Exception/SocketException.php]: Can you add some doc what this exception is used for.

extend `\RuntimeException`?

File: src/Bundler/Runner/SingleProcessRunner.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+ * Gives the file name as the first argument.
+ *
+ * Puts the content of the file on stdin, expects response on stdout.
+ */
+class SingleProcessRunner implements RunnerInterface
+{
+    private $config;
+    private $files;
+
+    public const BUILT_IN_FILES = [
```
Comment (yannickl88) [inline src/Bundler/Runner/SingleProcessRunner.php]: private?

File: src/Bundler/Runner/UnixSocketRunner.php [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+declare(strict_types=1);
+
+namespace Hostnet\Component\Resolver\Bundler\Runner;
+
+use Hostnet\Component\Resolver\Bundler\ContentItem;
+use Hostnet\Component\Resolver\Bundler\Runner\Exception\SocketException;
+use Hostnet\Component\Resolver\Config\ConfigInterface;
+use Hostnet\Component\Resolver\File;
+
+class UnixSocketRunner implements RunnerInterface
```
Comment (yannickl88) [inline src/Bundler/Runner/UnixSocketRunner.php]: doc

File: src/Bundler/Runner/UnixSocketRunner.php [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+    public function execute(string $type, ContentItem $item): string
+    {
+        $file_name = File::makeAbsolutePath($item->file->path, $this->config->getProjectRoot());
+        $start     = microtime(true);
+        $response  = '';
+
+        while (true) {
+            if (microtime(true) - $start > 30) {
+                throw new \RuntimeException('Timeout of 30s');
```
Comment (yannickl88) [inline src/Bundler/Runner/UnixSocketRunner.php]: can you add a specific exception for this?

File: src/Bundler/Runner/UnixSocketRunner.php [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+    private function startBuildProcess()
+    {
+        $node_js = $this->config->getNodeJsExecutable();
+        $cmd     = sprintf(
+            'nohup %s %s > /dev/null &',
+            $node_js->getBinary(),
+            implode(DIRECTORY_SEPARATOR, [__DIR__, '..', '..', 'Resources', 'build.js'])
+        );
+        `$cmd`;
```
Comment (yannickl88) [inline src/Bundler/Runner/UnixSocketRunner.php]: possible usecase for this `symfony/process` component? It is already used for the other runner.

File: src/Config/FileConfig.php [hunk 6]
```diff
@@ -195,4 +198,18 @@ public function getEventDispatcher(): EventDispatcherInterface
...
+
+    /**
+     * Which runner should we use for the transitions?
+     *
+     * @return RunnerInterface
+     */
+    public function getRunner(): RunnerInterface
+    {
+        if ($this->config_file_contents['enable-unix-socket'] ?? false) {
+            return new UnixSocketRunner($this);
```
Comment (yannickl88) [inline src/Config/FileConfig.php]: maybe check if we are on a unix system? Not sure how though...

File: src/Resources/build.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+    var chunkProcessor = new ChunkProcessor();
+
+    client.on('error', function (error) { console.error('woah, wtf', error)});
+
+    function sendReply(isSuccessful, message)
+    {
+        console.log('numberOfRequestsHandled', numberOfRequestsHandled);
+        var stopProcess = numberOfRequestsHandled >= 1000;
+        numberOfRequestsHandled++;
```
Comment (yannickl88) [inline src/Resources/build.js]: var stopProcess = numberOfRequestsHandled++ >= 1000;

**Comment by yannickl88** [comment]
Small side-note: How are we going to debug the server if it fails for some reason? Is there a log we can see, or is the output passed back?

File: src/Resources/build.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+var replySender = require("./reply-sender");
+var processor = require('./processor');
+
+process.umask(0o000);
+
+// This server listens on a Unix socket at process.argv[2]
+var numberOfRequestsHandled = 0,
+    unixServer = net.createServer(function (client) {
+
+    var chunkProcessor = new ChunkProcessor();
```
Comment (yannickl88) [inline src/Resources/build.js]: indentation is off, should it not be indented once more?

```js
var numberOfRequestsHandled = 0,
    unixServer = net.createServer(function (client) {
        var chunkProcessor = new ChunkProcessor();
        // ..
    });
```

File: src/Resources/build.js [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        if (stopProcess) {
+            callback = function () { unixServer.close(); };
+        }
+
+        replySender.sendReply(client, isSuccessful, stopProcess, message, callback);
+    }
+
+    client.on('readable', function() {
+
+        var chunk, bytes = chunkProcessor.protocolReceiveBytes;
```
Comment (yannickl88) [inline src/Resources/build.js]: `let chunk, bytes = ...`

File: src/Resources/chunk-processor.js [hunk 10]
```diff
@@ -0,0 +1,94 @@
+module.exports = function ()  {
```
Comment (yannickl88) [inline src/Resources/chunk-processor.js]: you care not very consistent with wrapping the JS in a closure or not...

Comment (nicoschoenmaker) [reply src/Resources/chunk-processor.js]: The other classes are defined as singletons. In either case no variable is leaked to global scope.

**Pull request merged**